[kloong.96] 방탈출 인증 관리 미션 리뷰 요청합니다.

Kloong1 opened the pull request:
안녕하세요 nesoy!
주말을 지나고 오랜만에 리뷰 요청을 하는 것 같네요.
지난 미션은 마지막 단계가 남아 있는데, 주말에 일정이 있어서 아직 마무리를 짓지 못했습니다 ㅜㅜ
일단 이번주 미션부터 제출을 하고, 지난주 미션은 최대한 빨리 완성해서 PR 올리려고 하는데 시간이 부족하면 패스할 수도 있을 것 같네요... ㅠㅠ

## 미션 설명
이번 미션은 지난주에 짰던 방탈출 프로그램에 **인증 기능을 추가하는 미션** 입니다.
사용자가 로그인 요청을 하면, 아이디와 비밀번호를 검증해서 DB에 있는 Member인지 확인한 후 맞으면 토큰을 발급해줍니다.
이후 사용자는 토큰을 가지고 자신의 정보를 조회하거나, 예약을 생성/취소하게 됩니다.

## 제가 작성한 코드
스켈레톤 코드가 주어진 상태에서 진행을 했는데, 말이 스켈레톤 코드지 사실상 대부분의 동작하는 코드가 주어져 있는 상황에서 시작을 했습니다. 도메인 객체와 DTO, DAO, 서비스, 컨트롤러, 그리고 테스트 코드가 주어졌고, 그 코드들을 수정하고 필요한 경우 추가하는 방식으로 진행했습니다.

- 사용자가 로그인 요청을 하면 아이디 비밀번호를 검증해서 토큰을 발급 -> auth 컨트롤러, auth 서비스, 해더에서 토큰 추출하는 기능 구현
- 사용자가 토큰으로 자신의 정보 조회 -> member 컨트롤러 기능 구현
- 사용자가 예약을 생성/취소 -> reservation 컨트롤러 기능 구현
- 예약 생성/취소 요청 시 토큰으로 사용자 정보 얻어냄 -> LoginMember, LoginMemberArgumentResolver
- 기타 사용자 정의 예외 적용 -> exception 패키지
- 예외를 잡아서 알맞은 응답으로 처리 -> ControllerAdvice 구현
- 테스트 코드 추가 및 수정

대충 집중적으로 기능을 추가 및 변경한 부분은 위와 같습니다. 나머지는 조금씩 건드린 정도인 것 같네요.

단위 테스트는 이번 과제에서 핵심적인 부분은 아닌 것 같아서 통합 테스트 위주로 작성해봤습니다.테스트 코드도 어느정도 주어져있기도 했구요.
 
## 리뷰해주셨으면 하는 부분들
위에서 언급한 제가 작성한 코드 부분을 중점적으로 봐주셨으면 합니다.

아무래도 제가 작성한 코드보다 이미 있는 코드량이 더 많아서 저도 정확하게 뭘 봐달라고 요청을 드리기가 어렵네요.
또 기존 코드가 지난주까지 배우고 피드백 해주셨던 부분들을 지키지 않은 곳(예를 들어 응답 dto가 따로 없는 등...)이 좀 있어서 어디부터 어디까지 리팩토링을 해야하는지 고민이 많이 되더라구요.
또 DB 테이블이나 도메인 객체도 좀 의아한 부분도 많이 있었구요.

근데 그런 부분들을 굳이 리팩토링 하기 시작하면 완전 새로 짜는 느낌이 될 것 같기도 하고, 이번 과제의 취지에도 맞지 않는 것 같아서 인증 관련 기능을 적용하는 데만 집중해 보았습니다. 해당 부분들을 위주로 리뷰해주시면 감사하겠습니다.

바쁘실텐데 신경써서 리뷰해주셔서 항상 감사합니다!

Kloong1 committed:
docs: 요구사항 및 기능 구현 목록 작성

Kloong1 committed:
feat: MemberRequest 디폴트 생성자 추가

Kloong1 committed:
feat: MemberDao의 findByUsername() 의 반환형을 Optional로 변경

Kloong1 committed:
feat: 사용자 정의 예외의 생성자 추가

Kloong1 committed:
feat: NotExistMemberException 추가

Kloong1 committed:
feat: NotCorrectPasswordException 추가

Kloong1 committed:
feat: 사용자 이메일과 비밀번호로 토큰 발급하는 서비스 로직 구현

Kloong1 committed:
feat: TokenRequest 디폴트 생성자 추가

Kloong1 committed:
feat: AuthController 로그인 요청 처리 기능 구현

Kloong1 committed:
feat: AuthController 예외 처리 기능 구현

Kloong1 committed:
feat: Authorization 헤더의 토큰을 추출하는 기능 구현

Kloong1 committed:
feat: InvalidAuthorizationTokenException 추가

Kloong1 committed:
feat: MemberService 의 findByUsername() 메소드 구현

Kloong1 committed:
feat: Member 디폴트 생성자 추가

Kloong1 committed:
feat: 토큰으로 자기 자신 조회 요청 처리 로직 구현

Kloong1 committed:
docs: 1단계 완성

Kloong1 committed:
docs: 2단계 요구사항 추가

Kloong1 committed:
feat: 도메인, dto, db 관련 변경

Kloong1 committed:
feat: AuthorizationTokenExtractor의 extract() 메소드 오버로딩

Kloong1 committed:
feat: LoginMember 어노테이션에 대한 ArgumentResolver 구현

Kloong1 committed:
fix: 생성자에서 넘겨받은 파라미터를 대입하지 않은 버그 수정

Kloong1 committed:
feat: ThemeRequest 디폴트 생성자 추가

Kloong1 committed:
feat: WebMvcConfigurer 구현 - ArgumentResolver 추가

Kloong1 committed:
feat: 예약 생성 요청 처리 기능 구현

Kloong1 committed:
feat: ReservationDao의 findById() 반환형을 Optional로 변경

Kloong1 committed:
feat: NotExistReservationException 추가

Kloong1 committed:
feat: NotQualifiedMemberException 추가

Kloong1 committed:
feat: Reservation 을 소유하는 Member를 확인하는 메소드 구현

Kloong1 committed:
feat: 예약을 삭제할 때 예약의 소유자 Member를 확인하는 로직 구현

Kloong1 committed:
feat: DuplicateReservationException 추가

Kloong1 committed:
feat: 중복 예약 생성시 DuplicateReservationException 던지게 변경

Kloong1 committed:
fix: ReservationDao SQL 구문의 빠진 컬럼 추가

Kloong1 committed:
feat: 예약의 소유자만 삭제 요청을 가능하게 로직 구현

Kloong1 committed:
refactor: exception 패키지 생성

Kloong1 committed:
refactor: 상수값을 static final로 변경

Kloong1 committed:
refactor: AuthorizationTokenExtractor 를 auth.util 패키지로 이동

Kloong1 committed:
refactor: 기타 리팩토링

Kloong1 committed:
feat: 중복 객체 예외를 DuplicateEntityException 예외로 통일

Kloong1 committed:
feat: 객체가 존재하지 않을 때 예외를 NotExistEntityException으로 통일

Kloong1 committed:
refactor: JwtTokenProvider를 auth.util 패키지로 이동

Kloong1 committed:
feat: 최상위 예외 RoomEscapeException 추가

Kloong1 committed:
feat: 다른 사용자의 예약을 삭제하려하면 401 코드 반환

Kloong1 committed:
refactor: auth 테스트 코드 패키지 구조 변경

Kloong1 committed:
feat: AuthControllerAdvice 적용

Kloong1 committed:
feat: MemberControllerAdvice 적용

Kloong1 committed:
test: MemberController 통합 테스트 코드 추가

Kloong1 committed:
feat: Member에 equals, hashCode 구현

Kloong1 committed:
feat: Reservation의 isCorrectMember()를 checkOwner()로 이름 변경

Kloong1 committed:
feat: ReservationControllerAdvice 적용

Kloong1 committed:
feat: ScheduleDao의 findById() 반환형을 Optional로 변경

Kloong1 committed:
feat: ThemeDao의 findById() 반환형을 Optional로 변경

Kloong1 committed:
feat: ScheduleControllerAdvice 적용

Kloong1 committed:
feat: ThemeControllerAdvice 적용

Kloong1 committed:
refactor: 의미 없는 테스트 코드 삭제

Kloong1 committed:
fix: ControllerAdvice의 패키지 범위 지정

Kloong1 committed:
fix: @DirtiesContext 어노테이션 적용

Kloong1 committed:
refactor: 테스트 코드 전체적으로 리팩토링

Kloong1 committed:
feat: 예외 에러 메시지 추가

Kloong1 committed:
feat: URI.create() 대신 UriComponentsBuilder 사용

NESOY commented on src/main/java/nextstep/auth/AuthService.java:
```diff
@@ -0,0 +1,31 @@
+                .orElseThrow(() -> new NotExistEntityException(
+                        "존재하지 않는 멤버입니다 - " + tokenRequest.getUsername()));
+        if(member.checkWrongPassword(tokenRequest.getPassword())) {
+            throw new NotCorrectPasswordException("잘못된 비밀번호입니다");
```
보안상 최대한 정보를 노출하지 않는게 좋은 방향으로 알고 있습니다.
혹시 **에러메시지**만 통일해보면 어떨까요? 
- `id가 존재하지 않거나 잘못된 비밀번호입니다`

NESOY commented on src/main/java/nextstep/auth/util/JwtTokenProvider.java:
```diff
@@ -11,29 +11,29 @@
 public class JwtTokenProvider {
-    private String secretKey = "learning-test-spring";
-    private long validityInMilliseconds = 3600000;
+    private static final String SECRET_KEY = "learning-test-spring";
+    private static final long VALIDITY_IN_MILLISECONDS = 3_600_000;
```
👍

NESOY commented on src/main/java/nextstep/support/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    public boolean supportsParameter(MethodParameter parameter) {
+        return parameter.hasParameterAnnotation(LoginMember.class);
```
Annotation & Member라는 클래스인 경우에만 동작하도록 하면 어떨까요? 🤔

NESOY commented on src/test/java/nextstep/theme/ThemeE2ETest.java:
```diff
@@ -38,7 +38,8 @@ public void showThemes() {
                 .extract();
-        assertThat(response.jsonPath().getList(".").size()).isEqualTo(1);
+
+        assertThat(response.jsonPath().getList(".")).hasSize(1);
```
👍

NESOY commented on src/test/java/nextstep/reservation/ReservationE2ETest.java:
```diff
@@ -23,12 +26,16 @@
     private Long scheduleId;
-    private Long memberId;
+
+    MemberRequest memberRequest1;
+    MemberRequest memberRequest2;
```
TIP : 나중에 시간이 여유가 되신다면 적용해보는것도 좋을거 같습니다. 
테스트 코드도 결국엔 **유지보수 비용**이 발생하기 때문에 최대한 구조로 만들기 위한 노력이 필요합니다 ㅠ

JUnit5에서도 ParameterResolver를 활용하면 테스트 매개변수로 원하는 객체를 주입받을 수 있습니다 :)
https://github.com/NESOY/junit5-example/blob/master/src/test/java/extension_model/ExtendWithTest.java
https://github.com/NESOY/junit5-example/blob/master/src/test/java/extension_model/RandomParametersExtension.java

NESOY commented on src/test/java/nextstep/member/MemberE2ETest.java:
```diff
@@ -0,0 +1,96 @@
+                .statusCode(HttpStatus.CREATED.value());
+    }
+
+    @DisplayName("자기 자신의 정보를 조회")
```
TIP : 같은 Context인 경우 Nested 형태로 테스트를 작성할 수도 있습니다. :)
https://johngrib.github.io/wiki/junit5-nested/#junit5의-nested를-사용해-계층-구조-테스트-코드를-작성하자

NESOY commented on src/main/java/nextstep/auth/util/AuthorizationTokenExtractor.java:
```diff
@@ -0,0 +1,27 @@
+        if (authorization == null) {
+            return Optional.empty();
+        }
+        if (authorization.toLowerCase().startsWith(BEARER_TYPE.toLowerCase())) {
```
👍

NESOY commented on src/main/java/nextstep/member/MemberController.java:
```diff
@@ -1,29 +1,42 @@
-        Long id = 1L;
-        Member member = memberService.findById(id);
+    public ResponseEntity me(HttpServletRequest request) {
+        String token = AuthorizationTokenExtractor.extract(request)
+                .orElseThrow(() -> new InvalidAuthorizationTokenException("유효하지 않은 토큰입니다"));
+        if (!jwtTokenProvider.validateToken(token)) {
+            throw new InvalidAuthorizationTokenException("유효하지 않은 토큰입니다 - " + token);
+        }
+        String userName = jwtTokenProvider.getPrincipal(token);
```
이 부분을 Controller보다는 spring interceptor 활용해보면 어떨까요?

NESOY reviewed (changes requested):
단계가 진행될수록 코드 양이 점점 많아지는 걸 보면서.. kloong의 노력이 많이 느껴집니다. 🥹
그래도 매 단계가 진행될수록 이전 리뷰 내용이 다 적용된 모습을 보면 대단하신거 같습니다 😁
인증 & 테스트 위주로 리뷰 진행했습니다. 테스트 관련 TIP들은 참고만 하시면 됩니다.

Kloong1 committed:
feat: ErrorCode 추가

Kloong1 committed:
feat: 에러 응답 바디에 담아줄 객체 추가

Kloong1 committed:
feat: 사용자 정의 예외에 ErrorCode 필드 추가

Kloong1 committed:
feat: ErrorCode에 에러 코드 추가

Kloong1 committed:
feat: 예외를 생성할 때 ErrorCode를 주입해주게 변경

Kloong1 committed:
refactor: 사용자 정의 예외의 사용하지 않는 생성자 제거

Kloong1 committed:
feat: RoomEscapeException 에 message, getMessage() 추가

Throwable이 상속해주는 detailMessage와 getMessage()를 상속받아서
사용하면 super()로 Throwable cause만 넘겨줬을 때 detailMessage를
cause.toString()으로 초기화한다. 에러 응답 시에 getMessage()로 에러
메시지도 바디에 넣어주도록 구현했는데 cause.toString()으로 초기화된 값이
출력되면 보안상 좋지 않으므로 이렇게 변경했다.

Kloong1 committed:
fix: 잘못된 에러 코드와 함께 에러를 던지던 버그 수정

Kloong1 committed:
feat: RoomEscapeException을 핸들링하는 글로벌 ControllerAdvice 추가

Kloong1 committed:
feat: LoginMemberArgumentResolver가 파라미터 타입도 검증하게 변경

Kloong1 committed:
feat: 자기 자신 조회 요청 시 arumentResolver로 인증하게 로직 변경

Kloong1 committed:
refactor: import문 정리

Kloong1 commented:
안녕하세요 네소이!
피드백 해주신 부분들 적용해보았습니다.

테스트 관련 내용도 고민하고 있던 부분인데, 설날 연휴동안 공부해서 다음 과제부터 적용할 수 있도록 해보겠습니다. 말씀하신대로 테스트 코드가 길어지니까 유지보수가 힘들더군요 ㅎㅎ...

에러메시지 역시 좀 문제가 있어 보이긴 했는데, 피드백 해 주신 김에 한번 완전히 뜯어 고쳐 보았습니다.

먼저 `ErrorCode` enum을 만들어서 에러 상황에 대한 HTTP 응답, 에러 코드, 에러 메시지를 정의했습니다.

그리고 로직 수행 중 사용자 정의 예외가 발생하면, 해당 예외에 상황에 맞는 `ErrorCode`를 주입해서 예외를 던졌습니다.
이 때 필요하면 추가적인 에러 메시지도 함께 주입할 수 있게 만들었습니다.

그리고 기존에 있던 `ControllerAdvice`를 모두 삭제하고, 사용자 정의 예외의 최상위 예외인 `RoomEscapeException`을 핸들링하는 `RoomEscapeExceptionHandler` 를 구현해서 `@RestControllerAdvice` 어노테이션을 붙여줬습니다.

`RoomEscapeExceptionHandler`는 잡은 예외에서 `ErrorCode`를 꺼낸 뒤, `ErrorCode`에 미리 정의 된 HTTP 응답을 반환하는 역할을 합니다. 또 에러 응답을 할 때 바디에 `ErrorCode`에 정의된 에러 코드와 에러 메시지도 담아서 반환하게 됩니다.

이런 식으로 클라이언트가 통일된 형태의 에러 응답을 받을 수 있도록 구현해 보았습니다.

항상 좋은 리뷰 감사합니다!

NESOY commented on src/main/java/nextstep/member/MemberController.java:
```diff
@@ -1,29 +1,33 @@
-    public ResponseEntity me() {
-        Long id = 1L;
-        Member member = memberService.findById(id);
+    public ResponseEntity me(@LoginMember Member member) {
```
TIP
위 방법도 좋지만 Annotation이 포함되지 않는 API인 경우 인증 프로세스를 놓칠 수 있을거 같아요.
그렇기 때문에 interceptor를 활용해서 대부분 API에 인증 프로세스가 적용되는 방법도 고려해보시면 좋을거 같습니다. 😁

NESOY reviewed (approved):
@Kloong1 고생 많으셨습니다. 🙇‍♂️ 
다음 단계에서 뵙겠습니다.
